provider/pkg/provider/remote/copyController.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/sftp"
+
+	p "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/infer"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+)
+
+// These are not required. They indicate to Go that Command implements the following interfaces.
+// If the function signature doesn't match or isn't implemented, we get nice compile time errors in this file.
+var _ = (infer.CustomResource[CopyInputs, CopyOutputs])((*Copy)(nil))
+var _ = (infer.CustomCheck[CopyInputs])((*Copy)(nil))
+var _ = (infer.CustomUpdate[CopyInputs, CopyOutputs])((*Copy)(nil))
+
+func (c *Copy) Check(ctx context.Context, urn string, oldInputs, newInputs resource.PropertyMap) (CopyInputs, []p.CheckFailure, error) {
+	var failures []p.CheckFailure
+
+	hasAsset := newInputs.HasValue("asset")
+	hasArchive := newInputs.HasValue("archive")
+
+	if hasAsset && hasArchive {
+		failures = append(failures, p.CheckFailure{
+			Property: "asset",
+			Reason:   "only one of asset or archive can be set",
+		})
+	}
+	if !hasAsset && !hasArchive {
+		failures = append(failures, p.CheckFailure{
+			Property: "asset",
+			Reason:   "either asset or archive must be set",
+		})
+	}
+
+	inputs, newFailures, err := infer.DefaultCheck[CopyInputs](newInputs)
+	if err != nil {
+		return inputs, failures, err
+	}
+	failures = append(failures, newFailures...)
```

Comment:
Regardless of if there is an error, we should merge `failures` and `newFailures`.

```suggestion
	inputs, newFailures, err := infer.DefaultCheck[CopyInputs](newInputs)
	failures = append(failures, newFailures...)
	if err != nil {
		return inputs, failures, err
	}
```